### ISSUE TITLE
Bug: `client: filesystem: write` from action should not run if action not chosen

### ISSUE THREAD

**Issue by jpadams** [dagger/dagger#2909]
User intended to invoke `actions.two` through `dagger do two` and was surprised that `actions.one` was triggered by the fact that a `client: filesystem: write:` used `actions.one`'s output.

They figured anything not related to the requested action would not run. In other words, they felt that anything related to `actions.one` would only execute if they asked for `dagger do one`.

https://discord.com/channels/707636530424053791/709803368318632038/1005489427797266516

```
package main

import (
  "dagger.io/dagger"
  "universe.dagger.io/docker"
  "dagger.io/dagger/core"
)

_#def
_#def: dagger.#Plan & {
  client: {
    filesystem: {
      folders: {
        read: {
          contents: dagger.#FS
        }
        write: {
          contents: actions.one.folders.output
        }
      }
    }
  }
  actions: {
    one: {
      build: docker.#Build & {
        steps: [docker.#Pull & {
          source: "alpine"
        }, docker.#Copy & {
          dest:     "/tmp/folders"
          contents: client.filesystem.folders.read.contents
        }, docker.#Run & {
          command: {
            name: "touch"
            args: ["/tmp/folders/test"]
          }
        }]
      }
      folders: core.#Subdir & {
        input: build.output.rootfs
        path:  "/tmp/folders"
      }
    }
    two: {
      build: docker.#Build & {
        steps: [docker.#Pull & {
          source: "alpine"
        }, docker.#Copy & {
          dest:     "/tmp/folders"
          contents: client.filesystem.folders.read.contents
        }]
      }
    }
  }
}

```

The "solution" at the moment is to use an environment variable or another action with a `bool` in it with default set to `false`. With an `if` guard around the `client: filesystem: write` one can use a `--with` on the command line to include the `client` bit in the `plan`.

```
dagger.#Plan & {
    client: {
      if actions.optional != false {
        filesystem: ....
      }
    }
    actions: {
        optional: bool | *false
    }
}
```

User asked if a file or FS tree under `client` could be set at optional or inactive by default like one can with `client: env`
package main

```
import (
    "dagger.io/dagger"
    "universe.dagger.io/docker"
)

dagger.#Plan & {
    client: env: {
        // load as a string, using a default if not defined
        REGISTRY_USER: string | *"_token_"
        // load as a secret, but don't fail if not defined
        REGISTRY_TOKEN?: dagger.#Secret
    }

    actions: pull: docker.#Pull & {
        source: "registry.example.com/image"
        if client.env.REGISTRY_TOKEN != _|_ {
            auth: {
                username: client.env.REGISTRY_USER
                secret:   client.env.REGISTRY_TOKEN
            }
        }
    }
}
```
from https://docs.dagger.io/1203/client#environment-variables

_Originally posted by @jpadams in https://github.com/dagger/dagger/issues/2470#issuecomment-1207444647_

**Comment by helderco** [comment]
Just to help understand what's happening, this is a unique condition where if two `read` and `write` actions are defined for the same `filesystem` path, a forced dependency is created to make sure _read_ always happens before _write_. This was to fix #1581:

https://github.com/dagger/dagger/blob/c3ac7a18e590cd715b4f3613bc5ca0344c09f506/pkg/dagger.io/dagger/plan.cue#L19-L23

You can go around it by choosing different `filesystem` keys and setting their `path`:

```cue
dagger.#Plan & {
	client: {
		filesystem: {
			source: read: {
				contents: dagger.#FS
				path:     "folders"
			}
			output: write: {
				contents: actions.one.folders.output
				path:     "folders"
			}
		}
	}
```

**Comment by helderco** [comment]
So it works as designed for one use case (target action doing both read and write), but is confusing in another (target action only doing one read/write, when defined as a pair).
